### PR TITLE
feat: wasm support

### LINKED ISSUE #96: WASM Support
What is the WASM support status on this lib?

What needs to be done?

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Add wasm support
- Close #96 

## Solution
- [x] make everything compile with `--target wasm32-unknown-unknown`. the biggest challenge is probably disabling transports.
- [x] implement a wasm compatible WS provider using `web_sys::WebSocket`
- [x] make sure all abi generated stuff is wasm compatible.

Afterwards we can look at how to shrink it etc.

### CODE REVIEW

**Comment by gakonst** [comment]
Nice start - we should also add a CI job for it

**Comment by mattsse** [comment]
Some parts work, however anything related to deploying and interacting hangs, could be due to misconfig on my end 🤷‍♂️ 

To test the minimal wasm test:
1. start ganache in separate terminal : `ganache-cli`
2. run wasm-pack (installation reuqired `cargo install wasm-pack`)
```
cd ethers-providers
wasm-pack test --chrome
```
go to http://127.0.0.1:8000/ and check browser console
or run in headless (only firefox works)
`wasm-pack test --headless --firefox`

**Review by gakonst — APPROVED** [review]
LGTM, follow ups from our discussion:
* Simplify the example, remove the need for webpack and make it as "vanilla" as possible
* NodeJS does not seem to be working right now. Unclear if it's a bug in how we do the WASM inside Rust, or if it's a config issue.

File: ethers-contract/ethers-contract-abigen/Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,10 @@ hex = { version = "0.4.2", default-features = false, features = ["std"] }
 reqwest = { version = "0.11.3", features = ["blocking"] }
 once_cell = { version = "1.8.0", default-features = false }
 cargo_metadata = "0.14.0"
+cfg-if = "1.0.0"
+
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+getrandom = { version = "0.2", features = ["js"] }
```
Comment (gakonst) [inline ethers-contract/ethers-contract-abigen/Cargo.toml:30]: This doesn't seem to be used anywhere?

Comment (gakonst) [inline ethers-contract/ethers-contract-abigen/Cargo.toml:30]: This doesn't seem to be used anywhere?

File: ethers-providers/Cargo.toml [hunk 2]
```diff
@@ -36,14 +36,27 @@ pin-project = { version = "1.0.7", default-features = false }
...
+ws_stream_wasm = "0.7"
+wasm-bindgen = "0.2"
+wasm-bindgen-futures = "0.4"
+web-sys = { version  = "0.3", features = ["console"] }
+# this is currently necessary for `wasm-timer::Delay` to work
+parking_lot = { version = "0.11", features = ["wasm-bindgen"] }
 
 [dev-dependencies]
-ethers = { version = "0.4.0", path = ".." }
+ethers = { version = "0.4.0", path = "..", features = ["legacy"] }
```
[the remark this answers is not in the document]
Comment (mattsse) [reply ethers-providers/Cargo.toml]: initially, I had the tests in this crate and just forgot to remove the legacy feature here

File: ethers-providers/Cargo.toml [hunk 3]
```diff
@@ -36,14 +36,27 @@ pin-project = { version = "1.0.7", default-features = false }
 tracing = { version = "0.1.26", default-features = false }
 tracing-futures = { version = "0.2.5", default-features = false, features = ["std-future"] }
 
+bytes = { version  = "1.0.1", default-features = false, optional = true }
+wasm-timer = "0.2"
```
[the remark this answers is not in the document]
Comment (mattsse) [reply ethers-providers/Cargo.toml]: yeah, moved it to wasm32 deps

File: ethers-contract/ethers-contract-abigen/Cargo.toml (hunk 1 above)
Comment (mattsse) [reply ethers-contract/ethers-contract-abigen/Cargo.toml:31]: you're right, we don't actually use this directly, however we need to set this for it to compile standalone:
https://docs.rs/getrandom/0.2.3/getrandom/#indirect-dependencies

**Review by gakonst — APPROVED** [review]

**Pull request merged**